# Genomes mapping

Project for creating a total list of genes for a bacteria.

Pipeline can be run by downloading repository and executing *main.py* file. Before that, please download all the required 
python packages:
* **numpy** - pipeline uses some arrays from numpy
* **matplotlib** - plots
* **seaborn** - plots
* **pandas** - working with tables
* **biopython** - for running BLAST and using NCBI api
* **tqdm** - progress bar.

Also, please open file constants.py and set your email adress to the EMAIL constant.

File main.py executes following scripts from *./functions*:
1. **set_enviroment.py**
2. **download_genomes.py**
3. **create_db.py**
4. **align_to_db.py**

All the steps, that take a lot of time have **tqdm** inside, so you can always see progress bar in your python console. 
*main.py* writes result of it's work to a file *./logs/pipeline.log*. All the steps off the pipeline described below in details.

## set_enviroment.py

Creates subdirectories needed for the other steps of the pipeline.

Directory tree after work of the pipeline will look like this (except for .py files):
```bash
    .
    ├── data - folder where all the genomes and output files will be saved
    │   ├── assembly_summary.txt - table with all organism assemblies from refseq
    │   ├── GCF_000005845.2_ASM584v2 - folder with files for that assembly
    │   │   ├── GCF_000005845.2_ASM584v2_db_alignment.csv - generated by align_to_db()
    │   │   ├── GCF_000005845.2_ASM584v2_feature_table.txt - from NCBI
    │   │   ├── GCF_000005845.2_ASM584v2_genomic.fna - from NCBI
    │   │   ├── GCF_000005845.2_ASM584v2_protein.faa - from NCBI
    │   │   └── GCF_000005845.2_ASM584v2_protein.gpff - from NCBI
    │   ├── GCF_000006665.1_ASM666v1
    │   ...
    ├── logs
    │   ├── alignment.log - log of align_to_database()
    │   ├── database.log - log of create_database()
    │   ├── downloading.log - log of download_genomes()
    │   └── pipeline.log - general log
    └── plots - folder with graphs
```

## download_genomes.py
Downloads *assembly_summary.txt* from NCBI ftp-site. For example, for Escherichia coli the file will be downloaded from 
[that page](ftp://ftp.ncbi.nlm.nih.gov/genomes/refseq/bacteria/Escherichia_coli/). If you want to use another organism, 
please set it's name to **ORGANISM** constant in *./constants.py*
